### PR TITLE
Proto_IkePacket.c: fix MD5, SHA-256, SHA-384 and SHA-512 implementation

### LINKED ISSUE #552: Latest Version of SoftEther VPN Server on x86_64 Ubuntu Breaks L2TP/IPSec Connections
Hello!

   The latest version of SoftEther VPN Server on x86_64 Ubuntu breaks connections initiated from Windows XP, which renders the system to be stuck at "Connecting to 1.2.3.4" instead of entering next phase of "Verifying user name and password...". This issue does not occur on Windows Vista and onwards. Can someone fix the issue?

* [ V ] Can you reproduce? -> Yes!
* [ V ] Are you running the latest version of SoftEtherVPN? -> Yes

**SoftEther version: Ver 4.27, Build 9668, beta
**Component:** Server
**Operating system:** Ubuntu 18.04 LTS
**Architecture:** 64-bit

**Processor:**: Intel(R) Core(TM) i7-3930K CPU @ 3.20GHz

### Description

Clients from Windows XP cannot establish connections to VPN server via L2TP/IPSec protocol.

**Expected behavior:**

Connections should be established successfully instead of hanging there.

**Actual behavior:**

Connections from Windows XP hang at "Connecting to W.X.Y.Z" instead of entering "Verifying user name and password..." phase.

### Steps to reproduce

1. Setup SoftEther VPN Server on x86_64 Ubuntu.
2. Establish L2TP/IPSec connections from Windows XP.
3. Windows XP cannot make connections to the server.
4. Try with Windows Vista and onwards to establish VPN tunnels via L2TP/IPSec.
5. Vista and onwards can reach the server and establish VPN tunnels via L2TP?IPSec successfully.

Hope to get your reply soon.

Thanks! Much appreciated!

Regards,

Andy

### LINKED ISSUE #688: L2TP/IPsec connection not working with recent SoftEther VPN Server versions running on Windows XP
* [ ] Can you reproduce? -> Yes.
* [ ] Are you running the latest version of SoftEtherVPN? -> Yes, I tried it too but the one that works for me is an earlier version.

**SoftEther version:**
softether-vpnserver_vpnbridge-v4.27-9668-beta-2018.05.29-windows-x86_x64-intel.exe 

**Component:** [Server, Client, Bridge, etc.] ->Server
**Operating system:** [Windows, Linux, BSD, macOS, etc.] -> Windows (XP SP3)
**Architecture:** [64 bit, 32 bit] -> 32 bit

[In case it's a computer with known specs, such as the Raspberry Pi, you can specify it omitting the details.]
**Processor:** [Specify brand and model. Example: AMD Ryzen 7 1800x]

Doesn't matter, tested on very different CPUs and machines.

### Description

[Description of the bug]

**Expected behavior:**
 To accept L2TP/IPSec PSK connections

**Actual behavior:**
[What actually happened]
The negotiation process does not pass

### Steps to reproduce

1. [First step] Install any version *HIGHER* (newer) than "softether-vpnserver_vpnbridge-v4.25-9656-rtm-2018.01.15-windows-x86_x64-intel.exe"
2. [Second step] Try to connect to it from Android device with L2TP/IPSec PSK (no other clients tested). Your attempt will fail. You can dig the exact error from log files on the server side. As far as I remember it was some kind of internal error. 
3. Uninstall current version and Install "softether-vpnserver_vpnbridge-v4.25-9656-rtm-2018.01.15-windows-x86_x64-intel.exe". 

4. Try to connect the same VPN, don't touch the settings at all again and you will succeed.

### PR DESCRIPTION
**Pull request by davidebeatrici**
Fixes #552.

#294 added SHA-256, SHA-384, and SHA-512 support to the protocol, but part of it was removed in faee11ff096e7950b006ef1ae8dc636b6fe0131e, because it caused a buffer over-read crash.

It also broke the MD5 implementation because the switch-case block doesn't handle the type anymore:
https://github.com/SoftEtherVPN/SoftEtherVPN/blob/8fdd224ba7f5676f99d74c7c63f60e046d07ec97/src/Cedar/Proto_IkePacket.c#L2982-L2996

This pull request fixes all the implementations and improves the `IkeHMac()` function by using the dedicated hashing functions.

---

SoftEther VPN Patch Acceptance Policy:
http://www.softether.org/5-download/src/9.patch

I choose option 1.

### CODE REVIEW

File: src/Cedar/Proto_IkePacket.c [hunk 1]
```diff
@@ -2963,108 +2963,42 @@ void IkeHash(IKE_HASH *h, void *dst, void *src, UINT size)
...
+
 	// Validate arguments
 	if (h == NULL || dst == NULL || (key == NULL && key_size != 0) || (data == NULL && data_size != 0))
 	{
 		return;
 	}
 
-	switch (h->HashId)
-	{
+	switch(h->HashId) {
```
[the remark this answers is not in the document]
Comment (chipitsine) [reply src/Cedar/Proto_IkePacket.c]: shall we add .clang-format with "BreakBeforeBraces: Stroustrup" ? (or any other value, I do not care personally)

and make travis-ci check the format

[the remark this answers is not in the document]
Comment (davidebeatrici) [reply src/Cedar/Proto_IkePacket.c]: > No idea if the coding style guidelines changed, but let’s keep the existing coding style for consistency?

Thank you, I wrote it like this by mistake.

> shall we add .clang-format with "BreakBeforeBraces: Stroustrup" ? (or any other value, I do not care personally)
>
> and make travis-ci check the format

I think that we should create the `clang-format` configuration according to the current code style, as even moving the brackets would require a lot of changes throughout the code.

File: src/Cedar/Proto_IkePacket.c [hunk 2]
```diff
@@ -2963,108 +2963,42 @@ void IkeHash(IKE_HASH *h, void *dst, void *src, UINT size)
...
 
-	if (h->HashId == IKE_HASH_SHA1_ID)
-	{
-		// Use special function (fast) in the case of SHA-1
-		HMacSha1(dst, key, key_size, data, data_size);
-		return;
-	}
-	else if (h->HashId == IKE_HASH_MD5_ID)
+	if (md == NULL)
 	{
```
[the remark this answers is not in the document]
Comment (davidebeatrici) [reply src/Cedar/Proto_IkePacket.c]: Done.

**Comment by slytimer** [comment]
Hi. It seems that it doesn't work again on windows xp. Build 9744

**Pull request merged**